march_gait_selection/src/march_gait_selection/transition_gait/transition_subgait.py

```diff
@@ -159,13 +158,53 @@ def _get_limits(robot, joint_name):
         raise TransitionError('Robot does not contain joint {joint}'.format(joint=joint_name))
 
     @staticmethod
-    def _validate_transition_gait(old_subgait, transition_subgait, new_subgait):
+    def _transition_setpoint(old_setpoint, new_setpoint, new_factor):
+        """Create a transition setpoint with the use of the old setpoint, new setpoint and transition factor."""
+        old_factor = 1.0 - new_factor
 
+        position = (old_setpoint.position * old_factor) + (new_setpoint.position * new_factor)
+        velocity = (old_setpoint.velocity * old_factor) + (new_setpoint.velocity * new_factor)
+
+        return Setpoint(new_setpoint.time, position, velocity)
+
+    @staticmethod
+    def _get_all_unique_timestamps(old_subgait, new_subgait):
+        """Get all the timestamps from the subgaits, eliminate double."""
+        timestamps = sorted(set(old_subgait.get_unique_timestamps() + new_subgait.get_unique_timestamps()))
+        return [round(timestamp, Setpoint.digits) for timestamp in timestamps]
+
+    @staticmethod
+    def _validate_transition_gait(old_subgait, transition_subgait, new_subgait):
+        """Validate the transition point by creating a gait object which checks the trajectories."""
         subgaits = [old_subgait, transition_subgait, new_subgait]
-        from_subgaits = ['start', old_subgait.subgait_name, transition_subgait.subgait_name, new_subgait.subgait_name]
+        from_subgaits = ['start', old_subgait.subgait_name, transition_subgait.subgait_name,
+                         new_subgait.subgait_name]
         to_subgaits = [old_subgait.subgait_name, transition_subgait.subgait_name, new_subgait.subgait_name, 'end']
 
         try:
             Gait('transition', subgaits, from_subgaits, to_subgaits)
         except Exception as error:
             TransitionError('Error when creating transition: {er}'.format(er=error))
+
+    @staticmethod
+    def _validate_transition_trajectory(old_subgait, transition_subgait, new_subgait):
+        """Validate if the calculated trajectory is within the given subgaits."""
+        for transition_joint in transition_subgait.joints:
+            old_joint = old_subgait.get_joint(transition_joint.name)
+            new_joint = new_subgait.get_joint(transition_joint.name)
+
+            for old_setpoint, transition_setpoint, new_setpoint in zip(old_joint, transition_joint, new_joint):
+                if old_setpoint.time != transition_setpoint.time or new_setpoint.time != transition_setpoint.time:
+                    raise TransitionError('The transition timestamps are not equal to the old and new subgait')
+
+                if old_setpoint.position < transition_setpoint.position:
+                    if not transition_setpoint.position <= new_setpoint.position:
```

Comment:
Shouldn't you also check that it is above old_setpoint.position? I think if old_postion = 0, new_position = 1 and transition_position = -100000, then this method will accept it as valid. (Same for the code below)

Reply:
I don't think this is true. if the transition_position is -100000 and old position is 0. Then the second if statement is called;
 ```suggestion
                if old_setpoint.position > transition_setpoint.position:
                    if not transition_setpoint.position >= new_setpoint.position:
                        raise TransitionError
```
If in the scenario u are describing the new setpoint is 1 then the transition error is raised because it is not between the two given setpoints.

The method is arranged as followed:
The transition position is always **between** the two given setpoints because otherwise the transition would not fit the pattern of the setpoints. Earlier in the code the startpoint and endpoint are validated using the **validate trajectory** in the gait class. This function purely validates if the setpoints is between the old-new or new-old.

Reply:
Ah never mind my brain errored for a moment there